src/StoreMeta.h

```diff
@@ -80,67 +64,96 @@ enum {
      * As STORE_META_STD except that the swap_file_sz is a 64-bit integer instead of 32-bit.
      */
-    STORE_META_STD_LFS,
+    STORE_META_STD_LFS = 9,
 
     // TODO: document this TLV type code
-    STORE_META_OBJSIZE,
+    STORE_META_OBJSIZE = 10
 
-    STORE_META_STOREURL,    /**< the Store-ID url, if different to the normal URL */
-    STORE_META_VARY_ID,     /**< Unique ID linking variants */
-    STORE_META_END
+    // When adding values, update Store::SwapMetaTypeMax.
+    // When removing values, check Store::IgnoredSwapMetaType() and friends.
 };
 
-/// \ingroup SwapStoreAPI
-class StoreMeta
+namespace Store {
+
+/// swap meta type ID written to or loaded from Store
+using RawSwapMetaType = char;
+
+/// maximum value of a recognized swap meta type
+const int SwapMetaTypeMax = 12; // Use "inline constexpr ..." with C++17.
+
+/// Store entries with longer swap metadata field values are not swapped out and
+/// are considered invalid when validating being-loaded metadata. This arbitrary
+/// limit protects code that adds individual swap metadata field sizes from
+/// overflowing and also prevents allocation of huge buffers when loading
+/// variable-length fields. Reevaluate this limit when increasing MAX_URL.
+const size_t SwapMetaFieldValueLengthMax = 64*1024;
+
+// TODO: Move to src/store/SwapMetaReading or similar, along with StoreMetaUnpacker.
+/// a swap metadata field inside the buffer given to StoreMetaUnpacker
+class SwapMetaView
 {
-protected:
-    StoreMeta() : length(-1), value(nullptr), next(nullptr) { }
-    StoreMeta(const StoreMeta &);
-    StoreMeta& operator=(const StoreMeta &);
+public:
+    /// ensures that our fixed-size field value has the given expected length
+    void checkExpectedLength(size_t) const;
 
 public:
-    static bool validType(char);
-    static int const MaximumTLVLength;
-    static int const MinimumTLVLength;
-    static StoreMeta *Factory(char type, size_t len, void const *value);
-    static StoreMeta **Add(StoreMeta **tail, StoreMeta *aNode);
-    static void FreeList(StoreMeta **head);
-
-    virtual char getType() const = 0;
-    virtual bool validLength(int) const;
-    virtual bool checkConsistency(StoreEntry *) const;
-    virtual ~StoreMeta() {}
-
-    int length;
-    void *value;
-    tlv *next;
+    /// A serialized rawLength-byte value (i.e. V in swap meta TLV field).
+    /// The value contents may be completely malformed/bogus.
+    /// Not meaningful when type is STORE_META_VOID.
+    const void *rawValue = nullptr;
+
+    /// The number of bytes in rawValue (i.e. L in swap meta TLV field).
+    /// This length may not match the length of valid fields of this type.
+    /// Not meaningful when type is STORE_META_VOID.
+    size_t rawLength = 0;
+
+    /// The serialized type (i.e. T in swap meta TLV field).
+    /// This type value may not match any named by SwapMetaType.
+    Store::RawSwapMetaType rawType = 0;
```

Comment:
more magic numbers
```suggestion
    Store::RawSwapMetaType rawType = SwapMetaType::begin_;
```

Reply:
Addressed in 40ccf5b.

----

> > Alex: We initialize it for sanity/clarity/debugging sake only.
 
> Amos: This statement invalidates all the previous arguments you make about why this variable had to be initialized.

I do not see any invalidation or contradiction. Moreover, those "previous arguments" were not about the necessity of initialization. They were about specific inappropriate initialization values being suggested in this change request.

> How? Initializing to an invalid value is only of use for **manual** analysis.

The problem is not that the proposed initialization _value_ is invalid. The problem is that its _type_ is wrong. Zero is the correct _value_ (used in past PR code, in current PR code, and in suggestions inside this change request), but we need to use the right _type_. Anonymous zero works OK, but using a named value of the right type is better, and that is what the code is using now.

>  If 0 is not a valid output from a "serialize(X)" it should not be used - ever. Better to leave the variable undefined.

Yes, 0 is not a valid serialize(X) output, but that does not mean zero should not be used to initialize a RawSwapMetaType data member or that the data member should be left uninitialized! RawSwapMetaType is not just the result of serialize(). Thinking of that raw type that way would lead to all sorts of bugs. RawSwapMetaType is the type of the `char` field we read from swap. It has no "impossible" value. All values are possible.

We could use Optional<RawSwapMetaType> for rawType if we needed to distinguish the two cases: not loaded and loaded zero. We do not need to make that distinction because rawType is always loaded (the object simply does not exist if we fail to load its rawType data member).

We could leave rawType without in-class initialization (because it is always initialized in the constructor), but I think that in this particular case such design would be slightly worse than the current code because the current code removes any doubts about rawType state on the (complex) way to its initialization in the constructor. It also provides symmetry with the lengthOrGarbage deserialization code -- we initialize that variable before deserializing it, for very similar reasons.

> I asked elsewhere for `STORE_META_VOID` to be defined as a valid (but not `Honoured`) type code (of value 0). This piece of logic is one of the reasons.

RawSwapMeta `STORE_META_VOID` has the wrong type for this RawSwapMetaType context. It also creates a wrong impression that `STORE_META_VOID` can be serialized into zero RawSwapMetaType when, in reality, `STORE_META_VOID` cannot be serialized at all.

In summary, there are several initialization options here, and the current code uses the best one:

* Do not initialize. No extra code (but a bit scary and inconsistent).
* Initialize with anonymous 0. Misses the existence of RawSwapMetaTypeBottom.
* Initialize with RawSwapMetaTypeBottom. Safe and consistent.
* Initialize with `STORE_META_VOID`. Uses the wrong type and sends the wrong signal.

Reply:
We still have disagreements here. But the current code looks fine to me, so to prevent those disagreements stalling things I am resolving this.